flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/RangeClauseBoundedOverProcessFunction.scala

```diff
@@ -0,0 +1,213 @@
+    out: Collector[Row]): Unit = {
+    // gets all window data from state for the calculation
+    val inputs: JList[Row] = dataState.get(timestamp)
+
+    if (null != inputs) {
+
+      var accumulators = accumulatorState.value
+      var dataListIndex = 0
+      var aggregatesIndex = 0
+
+      // initialize when first run or failover recovery per key
+      if (null == accumulators) {
+        accumulators = new Row(aggregates.length)
+        aggregatesIndex = 0
+        while (aggregatesIndex < aggregates.length) {
+          accumulators.setField(aggregatesIndex, aggregates(aggregatesIndex).createAccumulator())
+          aggregatesIndex += 1
+        }
+      }
+
+      // keep up timestamps of retract data
+      val retractTsList: JList[Long] = new JArrayList[Long]
+
+      val dataTimestampIt = dataState.keys.iterator
+      while (dataTimestampIt.hasNext) {
+        val dataTs: Long = dataTimestampIt.next()
+        val offset = timestamp - dataTs
+        if (offset > precedingOffset) {
+          val retractDataList = dataState.get(dataTs)
+          dataListIndex = 0
+          while (dataListIndex < retractDataList.size()) {
+            aggregatesIndex = 0
+            while (aggregatesIndex < aggregates.length) {
+              val accumulator = accumulators.getField(aggregatesIndex).asInstanceOf[Accumulator]
+              aggregates(aggregatesIndex)
+                .retract(accumulator, retractDataList.get(dataListIndex).getField(aggFields(aggregatesIndex)))
+              aggregatesIndex += 1
+            }
+            dataListIndex += 1
+            retractTsList.add(dataTs)
+          }
+        }
+      }
+
+      // remove the data that has been retracted
+      dataListIndex = 0
+      while (dataListIndex < retractTsList.size) {
+        dataState.remove(retractTsList.get(dataListIndex))
+        dataListIndex += 1
+      }
+
+      // copy forwarded fields to output row
+      aggregatesIndex = 0
+      while (aggregatesIndex < forwardedFieldCount) {
+        output.setField(aggregatesIndex, inputs.get(0).getField(aggregatesIndex))
+        aggregatesIndex += 1
+      }
+
+      dataListIndex = 0
+      while (dataListIndex < inputs.size()) {
+        // accumulate current row and set aggregate in output row
+        aggregatesIndex = 0
+        while (aggregatesIndex < aggregates.length) {
+          val index = forwardedFieldCount + aggregatesIndex
+          val accumulator = accumulators.getField(aggregatesIndex).asInstanceOf[Accumulator]
+          aggregates(aggregatesIndex).accumulate(accumulator, inputs.get(dataListIndex).getField(aggFields(aggregatesIndex)))
+          if (dataListIndex >= (inputs.size() - 1)) {
+            output.setField(index, aggregates(aggregatesIndex).getValue(accumulator))
+          }
+          aggregatesIndex += 1
+        }
+        dataListIndex += 1
+      }
+
+
+      dataListIndex = 0
+      while (dataListIndex < inputs.size()) {
+        out.collect(output)
```

Comment:
We have to copy the forwarded fields for each row in `inputs` because each row might be different. The aggregates can be reused, because they are all the same for all records with the same time.